docs/clients/dotnet/21.2/migration-to-gRPC.md

```diff
@@ -0,0 +1,599 @@
+
+Instead of doing paging like that in TCP client:
+
+```csharp
+public async Task<IEnumerable<object>> LoadEvents(string stream)
+{
+    const int pageSize = 4096;
+
+    var start  = 0;
+    var events = new List<object>();
+
+    do
+    {
+        var page = await _connection.ReadStreamEventsForwardAsync(
+            stream, start, pageSize, true
+        );
+
+        if (page.Status == SliceReadStatus.StreamNotFound)
+            throw new ArgumentOutOfRangeException(
+                nameof(stream), $"Stream '{stream}' was not found"
+            );
+
+        events.AddRange(
+            page.Events.Select(Deserialize)
+        );
+        if (page.IsEndOfStream) break;
+
+        start += pageSize;
+    } while (true);
+
+    return events;
+}
+```
+
+You can simplify that in the gRPC client into:
+
+```csharp
+public async Task<IEnumerable<object>> LoadEvents(string stream)
+{
+    await using var readResult = _client.ReadStreamAsync(
+        Direction.Forwards,
+        stream,
+        StreamPosition.Start
+    );
+
+    if(await readResult.ReadState != ReadState.Ok)
+        throw new ArgumentOutOfRangeException(
+            nameof(stream), $"Stream '{stream}' was not found"
+        );
+
+    return await readResult
+        .Select(Deserialize)
+        .ToListAsync();
+}
+```
+
+### Serialisation
+
+The gRPC client uses `ReadOnlyMemory<byte>` instead of `byte` array to make the events processing more efficient. To support that, you need to modify your deserialisation logic slightly:
+
+```csharp{4-5}
+  object Deserialize(this ResolvedEvent resolvedEvent)
+  {
+      var dataType = TypeMapper.GetType(resolvedEvent.Event.EventType);
+-     var jsonData = Encoding.UTF8.GetString(resolvedEvent.Event.Data);
++     var jsonData = Encoding.UTF8.GetString(resolvedEvent.Event.Data.Span);
+      var data = JsonConvert.DeserializeObject(jsonData, dataType);
+      return data;
+  }
+```
+
+## Built-in retries
+
+The gRPC client handles reconnections internally. But contrary to the TCP one, it does not have built-in retries for failed operations. It only does retries for the persistent subscriptions. If your codebase depends on them, you should wrap operations with your custom retry policy (e.g. using [Polly](https://github.com/App-vNext/Polly) library).
+
+If you were using the following configuration:
+
+```csharp
+var settings = ConnectionSettings.Create()
+    .FailOnNoServerResponse()
+    .KeepReconnecting()
+    .SetOperationTimeoutTo(TimeSpan.FromSeconds(5))
+    .LimitRetriesForOperationTo(7);
+
+var connection = 
+    EventStoreConnection.Create(
+        connectionString,
+	settings 
+);
+```
+
+You can replace it with a retry policy using Polly:
+
+```csharp
+public static class RetryScope
+{
+    private static readonly AsyncPolicyWrap RetryPolicy
+        = Policy
+            .Handle<Exception>()
```

Comment:
Not sure it's a good idea to retry everything. Like `StreamDeleted` is better not to retry.

Reply:
Added warning for retry policy definition in: https://github.com/EventStore/documentation/pull/486/commits/ffc35a21c4c86f91e92896e4d467dece8d636351.

Reply:
I also replaced the example of retry to reads instead of appends in https://github.com/EventStore/documentation/pull/486/commits/1225180ca1aafe17ff5cda49c18adb89d77d1e85.